utils/staking-miner/src/monitor.rs

```diff
@@ -0,0 +1,167 @@
+			("chain_subscribeNewHeads", "chain_unsubscribeNewHeads")
+		} else {
+			("chain_subscribeFinalizedHeads", "chain_unsubscribeFinalizedHeads")
+		};
+
+		log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", sub, unsub);
+		let mut subscription: Subscription<Header> = client
+			.subscribe(&sub, JsonRpcParams::NoParams, &unsub)
+			.await
+			.unwrap();
+
+		while let Some(now) = subscription.next().await.unwrap() {
+			let hash = now.hash();
+			log::debug!(target: LOG_TARGET, "new event at #{:?} ({:?})", now.number, hash);
+
+			// we prefer doing this check before fetching anything into a remote-ext.
+			if ensure_signed_phase::<Runtime, Block>(&client, hash).await.is_err() {
+				log::debug!(target: LOG_TARGET, "phase closed, not interested in this block at all.");
+				continue;
+			};
+
+			// NOTE: we don't check the score of any of the submitted solutions. If we submit a weak
+			// one, as long as we are valid, we will end up getting our deposit back, so not a big
+			// deal for now. Note that to avoid an unfeasible solution, we should make sure that we
+			// only start the process on a finalized snapshot. If the signed phase is long enough,
+			// this will not be a solution.
+
+			// grab an externalities without staking, just the election snapshot.
+			let mut ext = crate::create_election_ext::<Runtime, Block>(shared.uri.clone(), Some(hash), false).await?;
+
+			if ensure_no_previous_solution::<Runtime, Block>(&mut ext, &signer.account).await.is_err() {
+				log::debug!(target: LOG_TARGET, "We already have a solution in this phase, skipping.");
+				continue;
+			}
+
+			let (raw_solution, witness) = crate::mine_checked::<Runtime>(&mut ext, 50)?;
+			log::info!(target: LOG_TARGET, "mined solution with {:?}", &raw_solution.score);
+
+			let nonce = crate::get_account_info::<Runtime>(&client, &signer.account, Some(hash))
+				.await?
+				.map(|i| i.nonce)
+				.expect("signer account is checked to exist upon startup; it can only die if it \
+					transfers funds out of it, or get slashed. If it does not exist at this point, \
+					it is likely due to a bug, or the signer got slashed. Terminating."
+				);
+			let tip = 0 as Balance;
+			let period = Runtime::BlockHashCount::get()
+				.checked_next_power_of_two()
+				.map(|c| c / 2)
+				.unwrap_or(2) as u64;
+			let current_block = now.number
+				.saturated_into::<u64>();
+			let era = Era::mortal(period, current_block);
+			let extrinsic = ext.execute_with(|| create_uxt(raw_solution, witness, signer.clone(), nonce, tip, era));
+			let bytes = sp_core::Bytes(extrinsic.encode());
+
+			let mut tx_subscription: Subscription<
+				TransactionStatus<<Block as BlockT>::Hash, <Block as BlockT>::Hash>
+			> = client
+				.subscribe(&"author_submitAndWatchExtrinsic", params! { bytes }, "author_unwatchExtrinsic")
+				.await
+				.unwrap();
+
+			let _success = while let Some(status_update) = tx_subscription.next().await.unwrap() {
```

Comment:
Seems weird to have both `unwrap` and `while let` in the same line. Maybe

```suggestion
			let _success = while let Some(Some(status_update)) = tx_subscription.next().await {
```